src/Plugin_TEST.cc

```diff
@@ -0,0 +1,248 @@
+  plugin2.SetFilename("filename2");
+
+  // This is similar to what std::swap does except it uses std::move for each
+  // assignment
+  sdf::Plugin tmp = std::move(plugin);
+  plugin = plugin2;
+  plugin2 = tmp;
+
+  EXPECT_EQ("pluginname", plugin2.Name());
+  EXPECT_EQ("filename", plugin2.Filename());
+
+  EXPECT_EQ("pluginname2", plugin.Name());
+  EXPECT_EQ("filename2", plugin.Filename());
+}
+
+/////////////////////////////////////////////////
+TEST(DOMPlugin, Load)
+{
+  sdf::Plugin plugin;
+  sdf::Errors errors;
+
+  // Null sdf
+  errors = plugin.Load(nullptr);
+  ASSERT_EQ(1u, errors.size());
+  EXPECT_EQ(sdf::ErrorCode::ELEMENT_MISSING, errors[0].Code());
+
+  // Bad element name
+  sdf::ElementPtr sdf(new sdf::Element());
+  sdf->SetName("bad");
+  errors = plugin.Load(sdf);
+  ASSERT_EQ(1u, errors.size());
+  EXPECT_EQ(sdf::ErrorCode::ELEMENT_INCORRECT_TYPE, errors[0].Code());
+  EXPECT_NE(nullptr, plugin.Element());
+
+  sdf->SetName("plugin");
+
+  // Missing name and filename attribute
+  errors = plugin.Load(sdf);
+  ASSERT_EQ(2u, errors.size());
+  EXPECT_EQ(sdf::ErrorCode::ATTRIBUTE_MISSING, errors[0].Code());
+  EXPECT_EQ(sdf::ErrorCode::ATTRIBUTE_MISSING, errors[1].Code());
+
+  sdf->AddAttribute("name", "string", "__default__", true);
+  sdf->GetAttribute("name")->Set<std::string>("my-plugin-name");
+
+  // Now just missing filename
+  errors = plugin.Load(sdf);
+  ASSERT_EQ(1u, errors.size());
+  EXPECT_EQ(sdf::ErrorCode::ATTRIBUTE_MISSING, errors[0].Code());
+
+  sdf->AddAttribute("filename", "string", "__default__", true);
+  sdf->GetAttribute("filename")->Set<std::string>("filename.so");
+
+  // No errors.
+  errors = plugin.Load(sdf);
+  ASSERT_TRUE(errors.empty());
+}
+
+/////////////////////////////////////////////////
+TEST(DOMPlugin, LoadWithChildren)
+{
+  std::string pluginStr = R"(
+    <sdf version='1.9'>
+     <plugin filename='MinimalScene' name='3D View'>
+        <ignition-gui>
+          <title>3D View</title>
+          <property type='bool' key='showTitleBar'>false</property>
+          <property type='string' key='state'>docked</property>
+        </ignition-gui>
+        <engine>ogre</engine>
+        <scene>scene</scene>
+        <ambient_light>0.4 0.4 0.4</ambient_light>
+        <background_color>0.8 0.8 0.8</background_color>
+        <camera_pose>-6 0 6 0 0.5 0</camera_pose>
+      </plugin>
+    </sdf>)";
+
+  sdf::ElementPtr elem(new sdf::Element);
+  sdf::initFile("plugin.sdf", elem);
+  sdf::readString(pluginStr, elem);
+
+  sdf::Plugin plugin;
+  sdf::Errors errors;
+  errors = plugin.Load(elem);
+  ASSERT_EQ(0u, errors.size());
+
+  EXPECT_EQ("3D View", plugin.Name());
+  EXPECT_EQ("MinimalScene", plugin.Filename());
+
+  // The elements should be the same
+  EXPECT_EQ(elem->ToString(""), plugin.Element()->ToString(""));
+
+  sdf::ElementPtr toElem = plugin.ToElement();
+
+  // The elements should be the same
+  EXPECT_EQ(elem->ToString(""), toElem->ToString(""));
```

Comment:
Is there a way we can compare against `pluginStr` instead of trusting two `ToString` calls?

Reply:
Done in 1f3b675